drivers/smbus/smbus_stm32.c

```diff
@@ -0,0 +1,318 @@
+{
+	const struct smbus_stm32_config *config = dev->config;
+
+	return i2c_write(config->i2c_dev, &command, sizeof(command), periph_addr);
+}
+
+static int smbus_stm32_byte_read(const struct device *dev, uint16_t periph_addr, uint8_t *byte)
+{
+	const struct smbus_stm32_config *config = dev->config;
+
+	return i2c_read(config->i2c_dev, byte, sizeof(*byte), periph_addr);
+}
+
+static int smbus_stm32_byte_data_write(const struct device *dev, uint16_t periph_addr,
+				       uint8_t command, uint8_t byte)
+{
+	const struct smbus_stm32_config *config = dev->config;
+	uint8_t buffer[] = {
+		command,
+		byte,
+	};
+
+	return i2c_write(config->i2c_dev, buffer, ARRAY_SIZE(buffer), periph_addr);
+}
+
+static int smbus_stm32_byte_data_read(const struct device *dev, uint16_t periph_addr,
+				      uint8_t command, uint8_t *byte)
+{
+	const struct smbus_stm32_config *config = dev->config;
+
+	return i2c_write_read(config->i2c_dev, periph_addr, &command, sizeof(command), byte,
+			      sizeof(*byte));
+}
+
+static int smbus_stm32_word_data_write(const struct device *dev, uint16_t periph_addr,
+				       uint8_t command, uint16_t word)
+{
+	const struct smbus_stm32_config *config = dev->config;
+	uint8_t buffer[sizeof(command) + sizeof(word)];
+
+	buffer[0] = command;
+	sys_put_le16(word, buffer + 1);
+
+	return i2c_write(config->i2c_dev, buffer, ARRAY_SIZE(buffer), periph_addr);
+}
+
+static int smbus_stm32_word_data_read(const struct device *dev, uint16_t periph_addr,
+				      uint8_t command, uint16_t *word)
+{
+	const struct smbus_stm32_config *config = dev->config;
+	int result;
+
+	result = i2c_write_read(config->i2c_dev, periph_addr, &command, sizeof(command), word,
+			      sizeof(*word));
+	*word = sys_le16_to_cpu(*word);
+
+	return result;
+}
+
+static int smbus_stm32_pcall(const struct device *dev, uint16_t periph_addr, uint8_t command,
+			     uint16_t send_word, uint16_t *recv_word)
+{
+	const struct smbus_stm32_config *config = dev->config;
+	uint8_t buffer[sizeof(command) + sizeof(send_word)];
+	int result;
+
+	buffer[0] = command;
+	sys_put_le16(send_word, buffer + 1);
+
+	result = i2c_write_read(config->i2c_dev, periph_addr, buffer, ARRAY_SIZE(buffer), recv_word,
+			      sizeof(*recv_word));
+	*recv_word = sys_le16_to_cpu(*recv_word);
+
+	return result;
+}
+
+static int smbus_stm32_block_write(const struct device *dev, uint16_t periph_addr, uint8_t command,
+				   uint8_t count, uint8_t *buf)
+{
+	const struct smbus_stm32_config *config = dev->config;
+	struct i2c_msg messages[] = {
+		{
+			.buf = &command,
+			.len = sizeof(command),
+			.flags = 0,
+		},
+		{
+			.buf = buf,
+			.len = count,
+			.flags = 0,
+		},
+	};
+
+	return i2c_transfer(config->i2c_dev, messages, ARRAY_SIZE(messages), periph_addr);
+}
+
+static int smbus_stm32_block_read(const struct device *dev, uint16_t periph_addr, uint8_t command,
+				  uint8_t *count, uint8_t *buf)
+{
+	return -ENOTSUP;
```

Comment:
I think you can assign NULL to api  structure, another issue, should we return `-ENOSYS` ?

Reply:
Yes, there actually already is a NULL check in z_impl_smbus_block_read in place which will return ENOSYS. I will assign NULL to the api structure.

Reply:
Fixed this one, as well as other possibly not implemented API functions.